### PR TITLE
Allow parsing in skipped row for C engine

### LINKED ISSUE #10911: read_csv parse issue with newline in quoted items combined with skiprows
Now I don't know if this is known or the desired behaviour but when I try to read certain rows from a large file that uses "~" (tilde) as a quotechar and use skiprows at the same time, the parser screws up as follows:
Note: I use "" in the output even though that isn't shown, if I didn't the markup would become messed up - sorry...
```python
>>> pd.read_csv(StringIO.StringIO('a,b,c\r~a\n b~,~e\n d~,~f\n f~\r1,2,~12\n 13\n 14~'), quotechar="~", skiprows=range(1,2) )
     a                  b        c
   "b~"           "e\n d"  "f\n f"
1    2     "12\n 13\n 14"     NaN
```
while the output I wish to get would be in this artificial case:
```
      a      b                 c
0     1      2     "12\n 13\n 14"
```
it seems when skipping rows, the parser ignores custom quotation - which in this case is undesired from my point of view. 

EDIT: It might well be that in the quoted texts newlines are not always \n but sometimes also \r.

EDIT2 (31.8.):
The lineterminator fix fails as far as I can see with the following example:
```python
>>> a = StringIO.StringIO('Text,url\r~example\r sentence\r one~,url1\r~example\n sentence\n two~,url2')
>>> pd.read_csv(a, quotechar="~", skiprows=range(1,2), lineterminator='\r' )
                            Text        url
0                       sentence        NaN
1                         "one~"       url1
2     "example\n sentence\n two"       url2
```
The problem is that there is a "text"-column in the csv with html-formatted textblocks as content. However, there is no saying what kind of newline the creators of the html used originally and the textblocks stem from different sources.
I might also add that it respects the quoting perfectly if one does not use "skiprows".

versioninfo: 
```
python: 2.7.10.final.0
python-bits: 64
OS: Windows
OS-release: 7

pandas: 0.16.2
nose: None
Cython: None
numpy: 1.9.2
scipy: 0.16.0
statsmodels: None
IPython: 4.0.0
sphinx: None
patsy: None
dateutil: 2.4.2
pytz: 2015.4
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.4.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.4.0
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
```

### LINKED ISSUE #12775: read_csv return wrong dataframe when setting skiprows.
#### Code Sample, a copy-pastable example if possible

```python
>>> import pandas as pd
>>> from StringIO import StringIO
>>> data = """id,text,num_lines
1,"line 11
line 12",2
2,"line 21
line 22",2
3,"line 31",1"""

>>> pd.read_csv(StringIO(data))
Out[2]: 
   id              text  num_lines
0   1  'line 11\nline 12'          2
1   2  'line 21\nline 22'          2
2   3           'line 31'          1

>>> pd.read_csv(StringIO(data), skiprows=[1])
Out[3]: 
         id              text  num_lines
0  'line 12"'                 2        NaN
1         2  'line 21\nline 22'        2.0
2         3           'line 31'        1.0
...
```

#### Expected Output
```python
>>> pd.read_csv(StringIO(data), skiprows=[1])
Out[3]: 
   id              text  num_lines
0   2  'line 21\nline 22'          2
1   3           'line 31'          1
...
```

It should skip '1,"line 11\nline 12",2' instead skip    '1,"line 11' 
 
#### output of ``pd.show_versions()``

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.10.final.0
python-bits: 64
OS: Linux
OS-release: 4.2.3-300.fc23.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.0
nose: 1.3.7
pip: 8.1.1
setuptools: 18.0.1
Cython: None
numpy: 1.11.0
scipy: 0.14.1
statsmodels: 0.6.1
xarray: None
IPython: 3.2.1
sphinx: 1.2.3
patsy: 0.4.1
dateutil: 2.5.2
pytz: 2016.3
blosc: None
bottleneck: 0.6.0
tables: 3.2.2
numexpr: 2.4.6
matplotlib: 1.4.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None

### PR DESCRIPTION
**Pull request by gfyoung**
Changes behaviour of C engine parser so that parsing is done on skipped rows so that they
are properly skipped.

Closes #10911.
Closes #12775.

### CODE REVIEW

**Comment by gfyoung** [comment]
The issue with allowing parsing on skipped rows is that there is a performance hit since we're spending more time processing rows that don't really matter.  How significant is it though do you think?

**Comment by jreback** [comment]
this for sure needs performance testing.

**Comment by gfyoung** [comment]
@jreback : it's actually a little more general than that.  I am addressing all of the cases that occur in `tokenizer.c`, which goes beyond just newline characters in quoted fields bug.  The other option is to add new logic in the `SKIP_LINE` state to check for quotes but I wasn't sure about how it would handle quotes within quotes?

**Comment by jreback** [comment]
well, you can run the benchmarks in the test suite and see what shows up. prob need some addtl ones that hae embeeded terminators and such. I would special case thinks for a situation like this.

**Comment by gfyoung** [comment]
"special case thinks" ?

**Comment by jreback** [comment]
I think

**Comment by gfyoung** [comment]
@jreback : Added some additional unit tests that my current fixes should be robust against.  Let me know if there are any additional unit tests that should be included/written.

**Comment by jorisvandenbossche** [comment]
@gfyoung There is one specific asv bench for skiprows (io_bench/read_csv_skiprows), but you can eg run all tests that contain `read_csv`, or all tests in the io_bench.py file

**Comment by gfyoung** [comment]
@jorisvandenbossche : how do I exactly run only one of the benchmark tests (i.e. just `io_bench.py`)?  The `vb_suite/run_suite.py` file doesn't seem to have any way to configure which benchmarks to run.

**Comment by jorisvandenbossche** [comment]
See the contributing docs on performance testing: http://pandas-docs.github.io/pandas-docs-travis/contributing.html#running-the-performance-test-suite (if anything is not clear, please indicate so we can improve the explanation)

**Comment by gfyoung** [comment]
@jorisvandenbossche : Thank you!  That link was very helpful.  I was looking on the `github.io` page for `vbench`, which was not as detailed as this.

<b>@Everyone</b>: I ran the benchmark for `io_bench.csv_skiprows`.  `master` currently runs it in 18.62 ms, while my PR runs it in 26.76 ms.

**Comment by jreback** [comment]
pls run all of the csv benchmarks

**Comment by gfyoung** [comment]
|                                                                            | master (ms) | PR (ms) |
| ---------------------------------------------------------------|:----------------:| -----------:|
| time_frame_to_csv                                             | 220.30         | 227.45    |
| time_frame_to_csv2                                           | 377.33         | 364.85    |
| time_frame_to_csv_date_formatting                  | 17.01           | 15.51      |
| time_frame_to_csv_mixed                                  | 269.36        | 257.73    |
| time_read_csv_infer_datetime_format_custom  | 18.10          | 18.30      |
| time_read_csv_infer_datetime_format_iso8601 | 2.97            | 3.10        |
| time_read_csv_infer_datetime_format_ymd       | 3.30            | 3.21        |
| time_read_csv_skiprows                                     | 18.23          | 25.74      |
| time_read_csv_standard                                     | 14.95          | 15.65      |
| time_read_parse_dates_iso8601                        | 2.05            | 2.14        |
| time_write_csv_standard                                     | 58.79          | 59.01      |

As can be observed according to `asv`, the performance has not been significantly impacted by my changes AFAICT.

**Comment by gfyoung** [comment]
If there are no complaints about the performance or the changes (they pass currently against Travis, I added `[no skip]` to change commit message), this PR is ready to merge.

File: pandas/src/parser/tokenizer.c [hunk 1]
```diff
@@ -743,26 +753,9 @@ int tokenize_delimited(parser_t *self, size_t line_limit)
...
             // start of record
-            if (skip_this_line(self, self->file_lines)) {
-                self->state = SKIP_LINE;
-                if (c == '\n') {
-                    END_LINE();
-                }
-                break;
-            }
-            else if (c == '\n') {
+            if (c == '\n') {
```
Comment (jreback) [inline pandas/src/parser/tokenizer.c]: why is this not ``lineterminator``?

File: pandas/src/parser/tokenizer.c [hunk 2]
```diff
@@ -1387,13 +1355,7 @@ int tokenize_whitespace(parser_t *self, size_t line_limit)
...
         case START_RECORD:
             // start of record
-            if (skip_this_line(self, self->file_lines)) {
-                self->state = SKIP_LINE;
-                if (c == '\n') {
-                    END_LINE();
-                }
-                break;
-            } else  if (c == '\n') {
+            if (c == '\n') {
```
Comment (jreback) [inline pandas/src/parser/tokenizer.c]: same here

**Comment by jreback** [comment]
pls have a look thru any open csv parsing issues. you changed some things which may allow some of them to be fixed e.g. things related to lineterminator and such.

File: pandas/src/parser/tokenizer.c (hunk 1 above)
Comment (gfyoung) [inline pandas/src/parser/tokenizer.c]: Because the `lineterminator` is not custom?  It never was `lineterminator` AFAICT, as you can see here <a href="https://github.com/pydata/pandas/pull/12900/files#diff-3010e5592cbe0505f6b76a93d861972eL765">here</a>.

Comment (jreback) [inline pandas/src/parser/tokenizer.c]: I get that you changed the code, I am looking at why

File: pandas/src/parser/tokenizer.c (hunk 2 above)
Comment (gfyoung) [inline pandas/src/parser/tokenizer.c]: same answer as above

File: pandas/src/parser/tokenizer.c (hunk 1 above)
Comment (gfyoung) [inline pandas/src/parser/tokenizer.c]: There is no change here.  This function was always checking for `\n`.

Comment (jreback) [inline pandas/src/parser/tokenizer.c]: and why is that (and not lineterminator)

Comment (gfyoung) [inline pandas/src/parser/tokenizer.c]: Because of this logic <a href="https://github.com/gfyoung/pandas/blob/skiprows-newlines-patch/pandas/src/parser/tokenizer.c#L1818">here</a>

Comment (jreback) [inline pandas/src/parser/tokenizer.c]: I c, can you add some comments to that effect then.

File: pandas/src/parser/tokenizer.c (hunk 2 above)
Comment (gfyoung) [inline pandas/src/parser/tokenizer.c]: Actually, this one is a bug, which I logged <a href="https://github.com/pydata/pandas/issues/12912">here</a>.

File: pandas/src/parser/tokenizer.c (hunk 1 above)
Comment (gfyoung) [inline pandas/src/parser/tokenizer.c]: Done.

File: pandas/src/parser/tokenizer.c (hunk 2 above)
Comment (gfyoung) [inline pandas/src/parser/tokenizer.c]: @jreback : I'm going to leave this one alone for a separate PR, as I think there is a larger issue to be dealt with in terms of `delim_whitespace` and `lineterminator` per my comments in #12912.

**Comment by gfyoung** [comment]
I took a look through all of the issues, and I think the two that I mentioned so far are the only ones that can be closed with this PR.

**Comment by jreback** [comment]
https://github.com/pydata/pandas/search?q=csv+lineterminator&type=Issues&utf8=%E2%9C%93

pls have a look at these and see if any are relevant

**Comment by gfyoung** [comment]
@jreback : I looked through all of the open issues tagged with `CSV` before I made that comment above.

**Comment by gfyoung** [comment]
@jreback : if there is nothing else, I think this is ready to merge.

**Comment by gfyoung** [comment]
@jreback : if there is nothing else, I think this is ready to merge.

**Comment by jorisvandenbossche** [comment]
For the skiprows benchmark, there is a considerable slowdown, which is of course logical since there is now more parsing, and the exact slowdown of course depends on the number of skipped rows (this is just the consequence of wanting to solve the bug). 
But the question is a bit of we are OK with this performance hit for such a corner case. I can imagine people use `skiprows` exactly to have a faster parsing when they need only part of the file? 
However, I also can't think of a nice way around this (apart from yet another keyword ..).

**Comment by gfyoung** [comment]
@jorisvandenbossche : Yes, that slowdown seems to be an unfortunate necessity at this point for the sake of correctness.  As I mentioned earlier, the other option was to add more logic under the `SKIP_LINE` block, but attempts to do so were not robust against additional testing and only appeared to duplicate existing code.

**Comment by gfyoung** [comment]
@jorisvandenbossche , @jreback : I'm going to table this one for now until #12939 gets merged in.  It will be a lot easier to merge a fix for this with only one (hopefully) tokenizing function to work with.

I also realized that one potential issue with my current fix is that we will be penalizing people if the fields in the skipped row are malformed, which is not what we want IINM.  Thus, I might need to go back to the drawing board (even if the correctness > the performance as the PR currently stands) to see whether I can build in some other cases and not penalize.

**Comment by gfyoung** [comment]
I knew there had to be a better way.  Just needed some fresh eyes on the issue and managed to get the adding of a couple of new states to work.  Let's see what Travis thinks.

**Comment by jreback** [comment]
ok pls rerun the perf tests as well (and maybe run vs 0.18.0) as a comparison

**Comment by gfyoung** [comment]
PR Branch
========
<pre>io_bench.frame_to_csv.time_frame_to_csv           212.25ms
io_bench.frame_to_csv2.time_frame_to_csv2         356.49ms
...formatting.time_frame_to_csv_date_formatting    15.38ms
...h.frame_to_csv_mixed.time_frame_to_csv_mixed   253.26ms
...m.time_read_csv_infer_datetime_format_custom    18.20ms
...time_read_csv_infer_datetime_format_iso8601     3.00ms
..._ymd.time_read_csv_infer_datetime_format_ymd     3.21ms
...nch.read_csv_skiprows.time_read_csv_skiprows    20.18ms
...nch.read_csv_standard.time_read_csv_standard    16.00ms
..._dates_iso8601.time_read_parse_dates_iso8601     2.07ms
...h.write_csv_standard.time_write_csv_standard    59.00ms</pre>
Master Branch
===========
<pre>io_bench.frame_to_csv.time_frame_to_csv           216.67ms
io_bench.frame_to_csv2.time_frame_to_csv2         353.74ms
...formatting.time_frame_to_csv_date_formatting    15.28ms
...h.frame_to_csv_mixed.time_frame_to_csv_mixed   254.94ms
...m.time_read_csv_infer_datetime_format_custom    17.82ms
....time_read_csv_infer_datetime_format_iso8601     3.03ms
..._ymd.time_read_csv_infer_datetime_format_ymd     3.16ms
...nch.read_csv_skiprows.time_read_csv_skiprows    19.74ms
...nch.read_csv_standard.time_read_csv_standard    16.08ms
..._dates_iso8601.time_read_parse_dates_iso8601     2.11ms
...h.write_csv_standard.time_write_csv_standard    59.25ms</pre>

**Comment by gfyoung** [comment]
FYI: I think you have an old build from your most recent PR that you just merged running on Travis right now.

**Comment by gfyoung** [comment]
@jreback : Travis is giving the green light, and I don't see any major timing discrepancies.  Ready to merge if there is nothing else.

**Comment by gfyoung** [comment]
@jreback : Travis is giving the green light, and I don't see any major timing discrepancies.  Ready to merge if there is nothing else.

File: pandas/src/parser/tokenizer.h [hunk 3]
```diff
@@ -124,6 +124,8 @@ typedef enum {
     EAT_LINE_COMMENT,
     WHITESPACE_LINE,
     SKIP_LINE,
+    QUOTE_IN_SKIP_LINE,
```
Comment (jreback) [inline pandas/src/parser/tokenizer.h]: at some point maybe document what these states actually mean a bit more (I know its mostly self-explanatory, but for future readers)

**Comment by jreback** [comment]
thanks!
